File: backend/src/application/use-cases/search/RefreshProductsCatalogUseCase.ts
import type { TriggerManualScrapeUseCasePort } from "@application/ports/incoming/TriggerManualScrapeUseCasePort";
import type { ProductCatalogRepository } from "@application/ports/outgoing/ProductCatalogRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type { RefreshProductsCatalogInput } from "./contracts";

export class RefreshProductsCatalogUseCase {
	constructor(
		private readonly triggerManualScrapeHandler: TriggerManualScrapeUseCasePort,
		private readonly productCatalogRepository: ProductCatalogRepository,
		private readonly logger: LoggerPort,
	) {}

	async execute(input: RefreshProductsCatalogInput): Promise<void> {
		this.logger.info(
			`[RefreshProductsCatalogUseCase] Background refresh started for query: "${input.query}", postalCode: "${input.postalCode}"`,
		);
		const liveResult = await this.triggerManualScrapeHandler.execute({
			query: input.query,
			postalCode: input.postalCode,
		});
		const savedProductsCount = await this.productCatalogRepository.save(
			liveResult.results,
			input.postalCode,
		);

		this.logger.info(
			`[RefreshProductsCatalogUseCase] Background refresh completed for "${input.query}". Saved: ${savedProductsCount}. Warnings: ${liveResult.warnings.length}`,
		);
	}
}

File: backend-dotnet/src/api/Application/UseCases/Search/RefreshProductsCatalogUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;

namespace SupermarketPlanner.Api.Application.UseCases.Search;

public sealed class RefreshProductsCatalogUseCase(
    ITriggerManualScrapeUseCase triggerManualScrapeUseCase,
    IProductCatalogRepository productCatalogRepository,
    ILogger<RefreshProductsCatalogUseCase> logger)
{
    public Task<Either<DomainError, Unit>> Invoke(
        RefreshProductsCatalogInput input,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "[RefreshProductsCatalogUseCase] Background refresh started for query: \"{Query}\", postalCode: \"{PostalCode}\"",
            input.Query, input.PostalCode);

        return triggerManualScrapeUseCase
            .Invoke(new TriggerManualScrapeInput(input.Query, input.PostalCode), cancellationToken)
            .BindAsync(liveResult => SaveAndLog(input, liveResult, cancellationToken));
    }

    private Task<Either<DomainError, Unit>> SaveAndLog(
        RefreshProductsCatalogInput input,
        SearchResult liveResult,
        CancellationToken cancellationToken) =>
        productCatalogRepository.Save(liveResult.Results, input.PostalCode, cancellationToken)
            .MapAsync(saved =>
            {
                logger.LogInformation(
                    "[RefreshProductsCatalogUseCase] Background refresh completed for \"{Query}\". Saved: {Count}. Warnings: {WarningCount}",
                    input.Query, saved, liveResult.Warnings.Count);
                return Unit.Value;
            });
}
